src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -30,36 +30,127 @@ public ChunkedEncodingReadStream(HttpConnection connection) : base(connection)
+                    {
+                        break;
+                    }
+                    connectionBuffer = connectionBuffer.Slice(bytesConsumed); // reposition our view of the buffer
+
+                    // We found a chunk size starting the buffer.  If after that we can find a line ending and
+                    // validate that as a valid line, then we can consume the line; we need to end this
+                    // method with either at least 1 byte left in the current chunk or having fully consumed
+                    // the chunk, including ending terminator.
+                    int lfPos = connectionBuffer.IndexOf((byte)'\n');
+                    if (lfPos == -1)
+                    {
+                        break;
+                    }
+
+                    // Account for the possibiliy that line endings are either \r or \r\n
+                    int lineEnd = lfPos;
+                    if (lfPos > 0 && connectionBuffer[lfPos - 1] == '\r')
+                    {
+                        lineEnd--;
+                    }
+
+                    if (lineEnd != 0)
+                    {
+                        // Validate the remainder of the chunk line, throwing if it's invalid.
+                        ValidateChunkLineAfterSize(connectionBuffer.Slice(0, lineEnd));
+                    }
+
+                    // Mark the data thus far as having been consumed.
+                    _connection.ConsumeRemainingBuffer(bytesConsumed + lfPos + 1);
+                    _chunkBytesRemaining = chunkSize;
+                    connectionBuffer = _connection.RemainingBuffer; // reposition our view of the buffer
+
+                    if (connectionBuffer.Length == 0)
+                    {
+                        // If we consumed the chunk line and exhausted the buffer, we're done.
+                        break;
+                    }
+
+                    // Copy/consume data from the chunk
+                    Debug.Assert(_chunkBytesRemaining > 0, "Expected positive chunk bytes remaining");
+                    Debug.Assert(connectionBuffer.Length > 0, "Expected positive connection buffer length");
+                    int bytesToRead = Math.Min((int)Math.Min((ulong)destination.Length, _chunkBytesRemaining), connectionBuffer.Length);
+                    if ((ulong)bytesToRead == _chunkBytesRemaining &&
+                        (connectionBuffer.Length == bytesToRead || // no terminator buffered
+                         connectionBuffer.Length == bytesToRead + 1 && connectionBuffer[bytesToRead] != '\n')) // another char buffered, but it's not a line terminator
+                    {
+                        // If we're going to consume the full chunk, we'll also need to consume its terminator, so we want
+                        // to make sure its terminator is buffered, as we can't here issue an asynchronous read to pull in
+                        // additional bytes.
```

Comment:
Yeah, it is a little weird, but it seemed like the least bad option.  I didn't want to have to remember additional information about the state of our parsing; the current approach is pretty clean, and not having this would introduce new states in the implicit state machine.  I'll think about it again, though, and see if I can come up with something better.

Reply:
My instinct is that just adding a "_needChunkTerminator" flag may be the cleanest thing to do.  I agree re not introducing new states, but it's a relatively simple state.